### PR TITLE
Feature ncm proxy config

### PR DESCRIPTION
**Pull request by rkausch**
- Created JAXB-annotated objects to be persisted to preferences. 
- Updated NetworkConfigurationManager and impl to use these objects for storing configuration. 
- Configured to preserve user inputs for non-selected proxy types. 
- Updated to store system proxy exceptions and manual proxy exceptions separately.

### CODE REVIEW

**Comment by jkae** [comment]
`src/test/java/io/opensphere/core/net/NetworkConfigurationManagerImplTest.java` needs to be updated

File: open-sphere-base/core/src/main/java/io/opensphere/core/NetworkConfigurationManager.java [hunk 1]
```diff
@@ -17,33 +23,15 @@
      */
     void addChangeListener(NetworkConfigurationChangeListener listener);
 
-    /**
-     * Get the system proxy config URL if one has been set.
-     *
-     * @return The URL or {@code null}.
-     */
-    String getProxyConfigUrl();
+    ConfigurationType getSelectedProxyType();
```
Comment (bcleveland1022) [inline open-sphere-base/core/src/main/java/io/opensphere/core/NetworkConfigurationManager.java]: Comments

File: open-sphere-base/core/src/main/java/io/opensphere/core/NetworkConfigurationManager.java [hunk 2]
```diff
@@ -17,33 +23,15 @@
...
-    String getProxyConfigUrl();
+    ConfigurationType getSelectedProxyType();
 
-    /**
-     * Get the hosts to be excluded from proxy usage.
-     *
-     * @return The string comprising the host patterns.
-     */
-    String getProxyExclusions();
+    NoProxyConfiguration getNoProxyConfiguration();
```
Comment (bcleveland1022) [inline open-sphere-base/core/src/main/java/io/opensphere/core/NetworkConfigurationManager.java]: missing comments.

File: open-sphere-base/core/src/main/java/io/opensphere/core/NetworkConfigurationManager.java [hunk 3]
```diff
@@ -17,33 +23,15 @@
...
 
-    /**
-     * Get the system proxy host if one has been set.
-     *
-     * @return The host or {@code ""}.
-     */
-    String getProxyHost();
+    SystemProxyConfiguration getSystemConfiguration();
```
Comment (bcleveland1022) [inline open-sphere-base/core/src/main/java/io/opensphere/core/NetworkConfigurationManager.java]: Missing comments.

File: open-sphere-base/core/src/main/java/io/opensphere/core/NetworkConfigurationManager.java [hunk 4]
```diff
@@ -17,33 +23,15 @@
...
 
-    /**
-     * Get the system proxy port if one has been set.
-     *
-     * @return The port or -1.
-     */
-    int getProxyPort();
+    UrlProxyConfiguration getUrlConfiguration();
```
Comment (bcleveland1022) [inline open-sphere-base/core/src/main/java/io/opensphere/core/NetworkConfigurationManager.java]: Missing comments.

File: open-sphere-base/core/src/main/java/io/opensphere/core/NetworkConfigurationManager.java [hunk 5]
```diff
@@ -17,33 +23,15 @@
...
+
+    ManualProxyConfiguration getManualConfiguration();
```
Comment (bcleveland1022) [inline open-sphere-base/core/src/main/java/io/opensphere/core/NetworkConfigurationManager.java]: Missing comments.

File: open-sphere-base/core/src/main/java/io/opensphere/core/NetworkConfigurationManager.java [hunk 6]
```diff
@@ -71,15 +52,11 @@
...
-     *
-     * @param host the host, or {@code ""} to disable the proxy
-     * @param port the port number
-     * @param useSystemProxies {@code true} if system proxies should be used
-     * @param configUrl the config URL
-     * @param hostPatterns the string comprising the host patterns
+     * @param configurationType
      */
-    void setProxyConfiguration(String host, int port, boolean useSystemProxies, String configUrl, String hostPatterns);
+    void setSelectedProxyType(ConfigurationType configurationType);
```
Comment (bcleveland1022) [inline open-sphere-base/core/src/main/java/io/opensphere/core/NetworkConfigurationManager.java]: Does this have comments github's diff is weird.

File: open-sphere-base/core/src/main/java/io/opensphere/core/NetworkConfigurationManager.java [hunk 7]
```diff
@@ -71,15 +52,11 @@
...
+
+    void persistConfiguration();
```
Comment (bcleveland1022) [inline open-sphere-base/core/src/main/java/io/opensphere/core/NetworkConfigurationManager.java]: Comments.

File: open-sphere-base/core/src/main/java/io/opensphere/core/net/NetworkConfigurationManagerImpl.java [hunk 8]
```diff
@@ -61,102 +58,130 @@ public void addChangeListener(NetworkConfigurationChangeListener listener)
...
     }
 
     @Override
-    public String getProxyExclusions()
+    public void persistConfiguration()
     {
-        return myPrefs.getString(PROXY_EXCLUSION_PATTERNS_PREF_KEY, "");
+        synchronized (PROXY_CONFIG_LOCK)
+        {
+            myPreferences.putJAXBObject("configurations", myConfigurations, false, this);
```
Comment (bcleveland1022) [inline open-sphere-base/core/src/main/java/io/opensphere/core/net/NetworkConfigurationManagerImpl.java]: Is the saved file structure changing?  If so you will need to add migrators to convert the old config file to the correct format.  Also the config files that are packaged with the installer will need to be updated as well.

File: open-sphere-base/core/src/main/java/io/opensphere/core/net/NetworkConfigurationManagerImpl.java [hunk 9]
```diff
@@ -1,11 +1,18 @@
 package io.opensphere.core.net;
 
+import java.util.Collections;
```
Comment (bcleveland1022) [inline open-sphere-base/core/src/main/java/io/opensphere/core/net/NetworkConfigurationManagerImpl.java]: This class' test is failing.

File: open-sphere-base/core/src/main/java/io/opensphere/core/NetworkConfigurationManager.java (hunk 1 above)
Comment (rkausch) [reply open-sphere-base/core/src/main/java/io/opensphere/core/NetworkConfigurationManager.java]: Ack. How'd I miss that. Fixed.

File: open-sphere-base/core/src/main/java/io/opensphere/core/NetworkConfigurationManager.java (hunk 2 above)
Comment (rkausch) [reply open-sphere-base/core/src/main/java/io/opensphere/core/NetworkConfigurationManager.java]: Fuxed

File: open-sphere-base/core/src/main/java/io/opensphere/core/NetworkConfigurationManager.java (hunk 3 above)
Comment (rkausch) [reply open-sphere-base/core/src/main/java/io/opensphere/core/NetworkConfigurationManager.java]: Fixed

File: open-sphere-base/core/src/main/java/io/opensphere/core/NetworkConfigurationManager.java (hunk 5 above)
Comment (rkausch) [reply open-sphere-base/core/src/main/java/io/opensphere/core/NetworkConfigurationManager.java]: Fixed

File: open-sphere-base/core/src/main/java/io/opensphere/core/NetworkConfigurationManager.java (hunk 6 above)
Comment (rkausch) [reply open-sphere-base/core/src/main/java/io/opensphere/core/NetworkConfigurationManager.java]: It didn't, so, fixed.

File: open-sphere-base/core/src/main/java/io/opensphere/core/NetworkConfigurationManager.java (hunk 7 above)
Comment (rkausch) [reply open-sphere-base/core/src/main/java/io/opensphere/core/NetworkConfigurationManager.java]: Fixed

File: open-sphere-base/core/src/main/java/io/opensphere/core/net/NetworkConfigurationManagerImpl.java (hunk 8 above)
Comment (rkausch) [reply open-sphere-base/core/src/main/java/io/opensphere/core/net/NetworkConfigurationManagerImpl.java]: Yep. We're implementing the migrators as part of another ticket (5659). I'll update the config files in the installers.

**Pull request merged**